opamp-client/src/main/java/io/opentelemetry/opamp/client/internal/request/service/HttpRequestService.java

```diff
@@ -0,0 +1,225 @@
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import opamp.proto.AgentToServer;
+import opamp.proto.ServerErrorResponse;
+import opamp.proto.ServerErrorResponseType;
+import opamp.proto.ServerToAgent;
+
+public final class HttpRequestService implements RequestService, Runnable {
+  private final HttpSender requestSender;
+  private final PeriodicTaskExecutor executor;
+  private final PeriodicDelay periodicRequestDelay;
+  private final PeriodicDelay periodicRetryDelay;
+  private final AtomicBoolean retryModeEnabled = new AtomicBoolean(false);
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private final RetryAfterParser retryAfterParser;
+  @Nullable private Callback callback;
+  @Nullable private Supplier<Request> requestSupplier;
+  public static final PeriodicDelay DEFAULT_DELAY_BETWEEN_REQUESTS =
+      PeriodicDelay.ofFixedDuration(Duration.ofSeconds(30));
+
+  /**
+   * Creates an {@link HttpRequestService}.
+   *
+   * @param requestSender The HTTP sender implementation.
+   */
+  public static HttpRequestService create(HttpSender requestSender) {
+    return create(requestSender, DEFAULT_DELAY_BETWEEN_REQUESTS, DEFAULT_DELAY_BETWEEN_REQUESTS);
+  }
+
+  /**
+   * Creates an {@link HttpRequestService}.
+   *
+   * @param requestSender The HTTP sender implementation.
+   * @param periodicRequestDelay The time to wait between requests in general.
+   * @param periodicRetryDelay The time to wait between retries.
+   */
+  public static HttpRequestService create(
+      HttpSender requestSender,
+      PeriodicDelay periodicRequestDelay,
+      PeriodicDelay periodicRetryDelay) {
+    return new HttpRequestService(
+        requestSender,
+        PeriodicTaskExecutor.create(periodicRequestDelay),
+        periodicRequestDelay,
+        periodicRetryDelay,
+        RetryAfterParser.getInstance());
+  }
+
+  HttpRequestService(
+      HttpSender requestSender,
+      PeriodicTaskExecutor executor,
+      PeriodicDelay periodicRequestDelay,
+      PeriodicDelay periodicRetryDelay,
+      RetryAfterParser retryAfterParser) {
+    this.requestSender = requestSender;
+    this.executor = executor;
+    this.periodicRequestDelay = periodicRequestDelay;
+    this.periodicRetryDelay = periodicRetryDelay;
+    this.retryAfterParser = retryAfterParser;
+  }
+
+  @Override
+  public void start(Callback callback, Supplier<Request> requestSupplier) {
+    if (isRunning.compareAndSet(false, true)) {
+      this.callback = callback;
+      this.requestSupplier = requestSupplier;
+      executor.start(this);
+    } else {
+      throw new IllegalStateException("RequestDispatcher is already running");
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (isRunning.compareAndSet(true, false)) {
+      executor.executeNow();
+      executor.stop();
+    }
+  }
+
+  private void enableRetryMode(@Nullable Duration suggestedDelay) {
```

Comment:
Imo calling it `enableRetryMode` is misleading. The code never retries, it just sends the same request all the time (or you could say it always retries no matter whether the request succeeds or not). The only thing done here is updating the delay for the next request.
I need to think about this more, perhaps this can be simplified.

Reply:
Ah, got it. I thought it was meant as a full replacement. Ok let me take a deeper look to see how we can integrate both ideas such that we get the same behavior but in a clearer way (it's almost eod where I live so I'll probably add the code changes tomorrow morning).

Reply:
I looked a bit at the go implementation to get a better understanding what this is supposed to do at all. Although I don't know any go I feel it helped a lot.
I still think that you could improve the code.